// @technote-space/github-action-helper/dist/logger.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
/* eslint-disable @typescript-eslint/no-explicit-any */
const sprintf_js_1 = require("sprintf-js");
const core_1 = require("@actions/core");
/**
 * Logger
 */
class Logger {
    /**
     * @param {function|undefined} replacer replacer
     * @param {boolean} notUseGroup not use group?
     */
    constructor(replacer, notUseGroup = false) {
        this.notUseGroup = notUseGroup;
        /**
         * @param {string} message message
         * @return {string[]} messages
         */
        this.splitMessage = (message) => message.replace(/\r?\n$/, '').split(/\r?\n/);
        /**
         * @param {string} message message
         * @param {any[]} args args
         * @return {string} output string
         */
        this.getOutputString = (message, ...args) => sprintf_js_1.sprintf(this.replacer(message), ...args.map(arg => 'string' === typeof arg ? this.replacer(arg) : arg));
        /**
         * @param {function} output output function
         * @param {function|null} replacer replacer
         * @param {string|string[]} message message
         * @param {any[]} args args
         */
        this.multiLineOutput = (output, replacer, message, ...args) => {
            if ('string' !== typeof message) {
                message.forEach(message => {
                    this.multiLineOutput(output, replacer, message, ...args);
                });
                return;
            }
            this.splitMessage(message).forEach(message => output(this.getOutputString(replacer ? replacer(message) : message, ...args)));
        };
        /**
         * @param {string|string[]} message message
         * @param {any[]} args args
         * @return {void}
         */
        this.log = (message, ...args) => this.multiLineOutput(core_1.info, null, message, ...args);
        /**
         * @param {string|string[]} message message
         * @param {any[]} args args
         * @return {void}
         */
        this.info = (message, ...args) => this.multiLineOutput(core_1.info, message => `> ${message}`, message, ...args);
        /**
         * @param {string|string[]} message message
         * @param {any[]} args args
         * @return {void}
         */
        this.debug = (message, ...args) => this.multiLineOutput(core_1.debug, null, message, ...args);
        /**
         * @param {string|string[]} message message
         * @param {any[]} args args
         * @return {void}
         */
        this.error = (message, ...args) => this.multiLineOutput(core_1.error, null, message, ...args);
        /**
         * @param {string|string[]} message message
         * @param {any[]} args args
         * @return {void}
         */
        this.warn = (message, ...args) => this.multiLineOutput(core_1.warning, null, message, ...args);
        /**
         * @param {string|string[]} message message
         * @param {any[]} args args
         * @return {void}
         */
        this.displayCommand = (message, ...args) => this.multiLineOutput(core_1.info, message => `[command]${message}`, message, ...args);
        /**
         * @param {string|string[]} message message
         * @return {void}
         */
        this.displayStdout = (message) => this.multiLineOutput(core_1.info, message => `  >> ${message}`, message);
        /**
         * @param {string|string[]} message message
         * @return {void}
         */
        this.displayStderr = (message) => this.multiLineOutput(core_1.warning, message => `  >> ${message}`, message);
        /**
         * @param {string} message message
         * @param {any[]} args args
         * @return {void}
         */
        this.startProcess = (message, ...args) => {
            if (this.notUseGroup) {
                this.info(message, ...args);
                return;
            }
            this.endProcess();
            core_1.startGroup(this.getOutputString(message, ...args));
            Logger.isRequiredEndGroup = true;
        };
        /**
         * @return {void}
         */
        this.endProcess = () => {
            if (this.notUseGroup) {
                return;
            }
            if (Logger.isRequiredEndGroup) {
                core_1.endGroup();
                Logger.isRequiredEndGroup = false;
            }
        };
        this.replacer = replacer ? replacer : (text) => text;
    }
}
exports.default = Logger;
Logger.isRequiredEndGroup = false;
/**
 * @return {void}
 */
Logger.resetForTesting = () => {
    Logger.isRequiredEndGroup = false;
};
